app/upgrades.go

```diff
@@ -1,3 +1,274 @@
+		"fractional balances transferred from x/evmutil to x/precisebank",
+		"aggregate sum", aggregateSum,
+	)
+
+	remainder := InitializeRemainder(ctx, precisebankKeeper, aggregateSum)
+	logger.Info("remainder amount initialized in x/precisebank", "remainder", remainder)
+
+	// Migrate fractional balance reserve from x/evmutil to x/precisebank.
+	// This should be done **after** store migrations are completed in
+	// app.mm.RunMigrations, which migrates fractional balances to
+	// x/precisebank.
+	if err := TransferFractionalBalanceReserve(
+		ctx,
+		accountKeeper,
+		bankKeeper,
+		precisebankKeeper,
+	); err != nil {
+		return fmt.Errorf("reserve transfer: %w", err)
+	}
+
+	return nil
+}
+
+// TransferFractionalBalances migrates fractional balances from x/evmutil to
+// x/precisebank. It sets the fractional balance in x/precisebank and deletes
+// the account from x/evmutil. Returns the aggregate sum of all fractional
+// balances.
+func TransferFractionalBalances(
+	ctx sdk.Context,
+	evmutilKeeper evmutilkeeper.Keeper,
+	precisebankKeeper precisebankkeeper.Keeper,
+) (sdkmath.Int, error) {
+	aggregateSum := sdkmath.ZeroInt()
+
+	var iterErr error
+
+	evmutilKeeper.IterateAllAccounts(ctx, func(acc evmutiltypes.Account) bool {
+		// Set account balance in x/precisebank
+		precisebankKeeper.SetFractionalBalance(ctx, acc.Address, acc.Balance)
+
+		// Delete account from x/evmutil
+		iterErr := evmutilKeeper.SetAccount(ctx, evmutiltypes.Account{
+			Address: acc.Address,
+			// Set balance to 0 to delete it
+			Balance: sdkmath.ZeroInt(),
+		})
+
+		// Halt iteration if there was an error
+		if iterErr != nil {
+			return true
+		}
+
+		// Aggregate sum of all fractional balances
+		aggregateSum = aggregateSum.Add(acc.Balance)
+
+		// Continue iterating
+		return false
+	})
+
+	return aggregateSum, iterErr
+}
+
+// InitializeRemainder initializes the remainder amount in x/precisebank. It
+// calculates the remainder amount that is needed to ensure that the sum of all
+// fractional balances is a multiple of the conversion factor. The remainder
+// amount is stored in the store and returned.
+func InitializeRemainder(
+	ctx sdk.Context,
+	precisebankKeeper precisebankkeeper.Keeper,
+	aggregateSum sdkmath.Int,
+) sdkmath.Int {
+	// Extra fractional coins that exceed the conversion factor.
+	// This extra + remainder should equal the conversion factor to ensure
+	// (sum(fBalances) + remainder) % conversionFactor = 0
+	extraFractionalAmount := aggregateSum.Mod(precisebanktypes.ConversionFactor())
+	remainder := precisebanktypes.ConversionFactor().
+		Sub(extraFractionalAmount).
+		// Mod conversion factor to ensure remainder is valid.
+		// If extraFractionalAmount is a multiple of conversion factor, the
+		// remainder is 0.
+		Mod(precisebanktypes.ConversionFactor())
+
+	// Panics if the remainder is invalid. In a correct chain state and only
+	// mint/burns due to transfers, this would be 0.
+	precisebankKeeper.SetRemainderAmount(ctx, remainder)
```

Comment:
because we know this should be 0 on mainnet & testnet, can we add an assertion here to ensure that?

Reply:
I'm not 100% certain yet if it would be 0 if the evmutil bugs affected this - may need to double check how the state is in mirrornet or at least with a mainnet genesis export possibly

Reply:
[Added ticket to track](https://app.shortcut.com/kava-labs/story/13831/verify-mainnet-state-for-fractional-balance-migration)